brouhaha/inference.py

```diff
@@ -0,0 +1,174 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from einops import rearrange
+from pyannote.core import Segment, SlidingWindow, SlidingWindowFeature
+
+from pyannote.audio.core.model import Specifications
+from pyannote.audio.core.task import Resolution
+from pyannote.audio.utils.multi_task import map_with_specifications
+from pyannote.audio import Inference
+
+class BrouhahaInference(Inference):
+    def slide(
+        self,
+        waveform: torch.Tensor,
+        sample_rate: int,
+        hook: Optional[Callable],
+    ) -> Union[SlidingWindowFeature, Tuple[SlidingWindowFeature]]:
+        """Slide model on a waveform
+
+        Parameters
+        ----------
+        waveform: (num_channels, num_samples) torch.Tensor
+            Waveform.
+        sample_rate : int
+            Sample rate.
+        hook: Optional[Callable]
+            When a callable is provided, it is called everytime a batch is
+            processed with two keyword arguments:
+            - `completed`: the number of chunks that have been processed so far
+            - `total`: the total number of chunks
+
+        Returns
+        -------
+        output : (tuple of) SlidingWindowFeature
+            Model output. Shape is (num_chunks, dimension) for chunk-level tasks,
+            and (num_frames, dimension) for frame-level tasks.
+        """
+
+        window_size: int = self.model.audio.get_num_samples(self.duration)
+        step_size: int = round(self.step * sample_rate)
+        _, num_samples = waveform.shape
+
+        def __frames(
+            receptive_field, specifications: Optional[Specifications] = None
+        ) -> SlidingWindow:
+            if specifications.resolution == Resolution.CHUNK:
+                return SlidingWindow(start=0.0, duration=self.duration, step=self.step)
+            return receptive_field
+
+        frames: Union[SlidingWindow, Tuple[SlidingWindow]] = map_with_specifications(
+            self.model.specifications, __frames, self.model.receptive_field
+        )
+
+        # prepare complete chunks
+        if num_samples >= window_size:
+            chunks: torch.Tensor = rearrange(
+                waveform.unfold(1, window_size, step_size),
+                "channel chunk frame -> chunk channel frame",
+            )
+            num_chunks, _, _ = chunks.shape
+        else:
+            num_chunks = 0
+
+        # prepare last incomplete chunk
+        has_last_chunk = (num_samples < window_size) or (
+            num_samples - window_size
+        ) % step_size > 0
+
+        if has_last_chunk:
+            # repeat last chunk as many times necessary to get to window_size
+            last_chunk: torch.Tensor = waveform[:, num_chunks * step_size :]
+            channel, last_window_size = last_chunk.shape
+            num_repeat = window_size // last_window_size + 1
+            last_chunk = last_chunk.repeat((channel, num_repeat))
+            last_chunk = last_chunk[:, :window_size]
```

Comment:
Instead of the fix we've discussed offline, I'd rather repeat the last chunk to get to window_size instead of simply doing:
`last_chunk: torch.Tensor = waveform[:, -window_size :]`